test/model/recent_senders_test.dart

```diff
@@ -0,0 +1,153 @@
+import 'package:checks/checks.dart';
+import 'package:test/scaffolding.dart';
+import 'package:zulip/model/recent_senders.dart';
+import '../example_data.dart' as eg;
+
+void main() {
+  group('MessageIdTracker', () {
+    late MessageIdTracker idTracker;
+
+    void prepare() {
+      idTracker = MessageIdTracker();
+    }
+
+    test('starts with no ids', () {
+      prepare();
+      check(idTracker.debugIds).isEmpty();
+    });
+
+    test('calling add(id) adds the same id to the tracker just one time', () {
+      prepare();
+      check(idTracker.debugIds).isEmpty();
+      idTracker.add(1);
+      idTracker.add(1);
+      check(idTracker.debugIds.singleOrNull).equals(1);
+    });
+
+    test('ids are sorted ascendingly, with maxId pointing to the last element', () {
+      prepare();
+      check(idTracker.debugIds).isEmpty();
+      idTracker.add(1);
+      idTracker.add(9);
+      idTracker.add(0);
+      idTracker.add(5);
+      check(idTracker.debugIds).deepEquals([0, 1, 5, 9]);
+      check(idTracker.maxId).equals(idTracker.debugIds.last);
```

Comment:
I'd rather write this test in terms of the class's interface, the same interface that the app code needs from it. So that means checking `maxId` instead of peeking at the full list.

This is actually what prompted me to write the comment just above about the absence of `remove` — from the class's actual interface, there's no way to tell that 0, 1, and 5 are still in the list at the end, or that they're in order. Because there's really no way to tell, we don't actually *want* a test that tries to check that — it's not the test's business, or the business of any code but the implementation of the class itself.

For background on that approach, see this section of Zulip's "testing philosophy" doc:
https://zulip.readthedocs.io/en/latest/testing/philosophy.html#integration-testing-or-unit-testing

(OTOH after adding `remove`, then the full content of this test case can be had by removing 9 and checking that maxId is now 5, etc.)

Reply:
A further step in that direction is that this whole `MessageIdTracker` class can be taken as an internal implementation detail of `RecentSenders`, and all these tests can be written in terms of `RecentSenders`. Essentially they're all just examples of test cases one can write for `RecentSenders` that involve only a single sender in a single conversation.

That doesn't require them to get any less simple than they are in this version. I'll make separate comments below on the RecentSenders test cases about ways to simplify those.